digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java

```diff
@@ -1745,6 +1759,18 @@ StoredSession upsertAndLockSession(@Bind("projectId") int projectId,
                 " for update skip locked")
         Long lockTaskIfNotLocked(@Bind("id") long taskId);
 
+        @SqlUpdate("update tasks" +
+                " set updated_at = now(), retry_at = NULL, state = " + TaskStateCode.READY_CODE +
+                " from (" +
+                  "select id from tasks" +
+                  " where state in (" + TaskStateCode.RETRY_WAITING_CODE +"," + TaskStateCode.GROUP_RETRY_WAITING_CODE + ")" +
+                  " and retry_at \\<= now()" +
+                  " order by id for update" +
+                " ) lck" +
+                " where tasks.id = lck.id"
```

Comment:
IIRC the purpose of the sub query is fixing the order of lock of the rows.
If so, is there any formal documentation description on it ?
I have a minor concern whether it depends on the specific Pg version or undocumented internal behavior or not.

Reply:
As far as I know, the official document explicitly doesn't mention this issue and behavior. But we can find some explanations about the issue.

> UPDATE deals with rows in order they are fetched from a heap. In a common case it can be considered as unordered.

https://www.postgresql.org/message-id/CAKOSWNkb3Zy_YFQzwyRw3MRrU10LrMj04%2BHdByfQu6M1S5B7mg%40mail.gmail.com

And I think you can easily find this workaround (e.g. in the above post or https://stackoverflow.com/a/27263824/5101185) although I can't 100% guarantee this behavior won't change forever.